source/azure_iot_hub_client.c

```diff
@@ -0,0 +1,538 @@
+    AzureIoTHubClientResult_t ret;
+    AzureIoTMQTTStatus_t xResult;
+    bool xSessionPresent;
+    uint32_t bytesCopied = 0;
+    size_t mqtt_user_name_length;
+    az_result res;
+
+    if( xAzureIoTHubClientHandle == NULL )
+    {
+        AZLogError( ( "Failed to connect: Invalid argument \r\n" ) );
+        ret = eAzureIoTHubClientInvalidArgument;
+    }
+    else
+    {
+        /* Use scratch buffer for username/password */
+        xConnectInfo.pUserName = ( const char * ) xAzureIoTHubClientHandle->_internal.pucAzureIoTHubClientWorkingBuffer;
+        xConnectInfo.pPassword =  xConnectInfo.pUserName + azureiotUSERNAME_MAX;
+
+        if( az_result_failed( res = az_iot_hub_client_get_user_name( &xAzureIoTHubClientHandle->_internal.xAzureIoTHubClientCore,
+                                                                     ( char * ) xConnectInfo.pUserName, azureiotUSERNAME_MAX,
+                                                                     &mqtt_user_name_length ) ) )
+        {
+            AZLogError( ( "Failed to get username: %08x\r\n", res ) );
+            ret = eAzureIoTHubClientFailed;
+        }
+        /* Check if token refresh is set, then generate password */
+        else if( ( xAzureIoTHubClientHandle->_internal.pxAzureIoTHubClientTokenRefresh ) &&
+                 ( xAzureIoTHubClientHandle->_internal.pxAzureIoTHubClientTokenRefresh( xAzureIoTHubClientHandle,
+                                                                                           xAzureIoTHubClientHandle->_internal.xAzureIoTHubClientTimeFunction() + azureiothubDEFAULT_TOKEN_TIMEOUT_IN_SEC,
+                                                                                           xAzureIoTHubClientHandle->_internal.pucAzureIoTHubClientSymmetricKey,
+                                                                                           xAzureIoTHubClientHandle->_internal.ulAzureIoTHubClientSymmetricKeyLength,
+                                                                                           ( uint8_t * ) xConnectInfo.pPassword, azureiotPASSWORD_MAX,
+                                                                                           &bytesCopied ) ) )
+        {
+            AZLogError( ( "Failed to generate auth token \r\n" ) );
+            ret = eAzureIoTHubClientFailed;
+        }
+        else
+        {
+            xConnectInfo.cleanSession = cleanSession;
+            xConnectInfo.pClientIdentifier = ( const char * ) xAzureIoTHubClientHandle->_internal.pucDeviceID;
+            xConnectInfo.clientIdentifierLength = ( uint16_t ) xAzureIoTHubClientHandle->_internal.ulDeviceIDLength;
+            xConnectInfo.userNameLength = ( uint16_t ) mqtt_user_name_length;
+            xConnectInfo.keepAliveSeconds = azureiothubKEEP_ALIVE_TIMEOUT_SECONDS;
+            xConnectInfo.passwordLength = ( uint16_t ) bytesCopied;
+
+            /* Send MQTT CONNECT packet to broker. LWT is not used in this demo, so it
```

Comment:
1 - If we need this, please expand out LWT to Last Will and Testament.  I get it should be obvious but there's so many acronyms out there...
2 - This isn't a demo anymore!